pinot-common/src/main/java/org/apache/pinot/sql/parsers/rewriter/CLPDecodeRewriter.java

```diff
@@ -0,0 +1,176 @@
+ * names of all the columns in the group.
+ * <p>
+ * Usage:
+ * <pre>
+ *   clpDecode("columnGroupName"[, defaultValue])
+ * </pre>
+ * which will be rewritten to:
+ * <pre>
+ *   clpDecode("columnGroupName_logtype", "columnGroupName_dictionaryVars", "columnGroupName_encodedVars"[,
+ *   defaultValue])
+ * </pre>
+ * The "defaultValue" is optional. See
+ * {@link org.apache.pinot.core.operator.transform.function.CLPDecodeTransformFunction} for its description.
+ * <p>
+ * Sample queries:
+ * <pre>
+ *   SELECT clpDecode("message") FROM table
+ *   SELECT clpDecode("message", 'null') FROM table
+ * </pre>
+ * See {@link org.apache.pinot.core.operator.transform.function.CLPDecodeTransformFunction} for details about the
+ * underlying clpDecode transformer.
+ */
+public class CLPDecodeRewriter implements QueryRewriter {
+  public static final String LOGTYPE_COLUMN_SUFFIX = "_logtype";
+  public static final String DICTIONARY_VARS_COLUMN_SUFFIX = "_dictionaryVars";
+  public static final String ENCODED_VARS_COLUMN_SUFFIX = "_encodedVars";
+
+  private static final String _CLPDECODE_LOWERCASE_TRANSFORM_NAME =
+      TransformFunctionType.CLPDECODE.getName().toLowerCase();
+
+  @Override
+  public PinotQuery rewrite(PinotQuery pinotQuery) {
+    List<Expression> selectExpressions = pinotQuery.getSelectList();
+    if (null != selectExpressions) {
+      for (Expression e : selectExpressions) {
+        tryRewritingExpression(e);
+      }
+    }
+    List<Expression> groupByExpressions = pinotQuery.getGroupByList();
+    if (null != groupByExpressions) {
+      for (Expression e : groupByExpressions) {
+        tryRewritingExpression(e);
+      }
+    }
+    List<Expression> orderByExpressions = pinotQuery.getOrderByList();
+    if (null != orderByExpressions) {
+      for (Expression e : orderByExpressions) {
+        tryRewritingExpression(e);
+      }
+    }
+    tryRewritingExpression(pinotQuery.getFilterExpression());
+    tryRewritingExpression(pinotQuery.getHavingExpression());
+    return pinotQuery;
+  }
+
+  /**
+   * Rewrites any instances of clpDecode in the given expression
+   * @param expression Expression which may contain instances of clpDecode
+   */
+  private void tryRewritingExpression(Expression expression) {
+    if (null == expression) {
+      return;
+    }
+    Function function = expression.getFunctionCall();
+    if (null == function) {
+      return;
+    }
+
+    String functionName = function.getOperator();
+    if (functionName.equals(_CLPDECODE_LOWERCASE_TRANSFORM_NAME)) {
+      rewriteCLPDecodeFunction(expression);
+    } else {
+      // Function isn't a CLP function that needs rewriting, but the arguments might be, so we recursively process them.
+      for (Expression op : function.getOperands()) {
+        tryRewritingExpression(op);
+      }
+    }
+  }
+
+  /**
+   * Rewrites the given instance of clpDecode as described in the class docstring
```

Comment:
what is docstring? typo?

Reply:
Ah, I think it's a [general term](https://en.wikipedia.org/wiki/Docstring) for API documentation comments, but the more specific term in Java is a Javadoc, so switched to that.